File: MMMLPDS/VectorAdd/vec_add_mpi.cpp
#include <mpi.h>
#include <iostream>
#include <vector>

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);

    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    double t_start = MPI_Wtime();

    const int N = 1000000;
    int local_n = N / size;

    std::vector<float> A(local_n), B(local_n), C(local_n);
    std::vector<float> fullA, fullB, fullC;

    if (rank == 0) {
        fullA.resize(N, 1.0f);
        fullB.resize(N, 1.0f);
    }

    MPI_Scatter(fullA.data(), local_n, MPI_FLOAT, A.data(), local_n, MPI_FLOAT, 0, MPI_COMM_WORLD);
    MPI_Scatter(fullB.data(), local_n, MPI_FLOAT, B.data(), local_n, MPI_FLOAT, 0, MPI_COMM_WORLD);

    for (int i = 0; i < local_n; ++i)
        C[i] = A[i] + B[i];

    if (rank == 0) fullC.resize(N);
    MPI_Gather(C.data(), local_n, MPI_FLOAT, fullC.data(), local_n, MPI_FLOAT, 0, MPI_COMM_WORLD);

    double t_end = MPI_Wtime();

    if (rank == 0) {
        std::cout << "Total runtime (MPI, including scatter/gather): " << (t_end - t_start) << " s" << std::endl;
        std::cout << "C[0] = " << fullC[0] << ", C[N-1] = " << fullC[N-1] << std::endl;
    }

    MPI_Finalize();
}

File: MMMLPDS/VectorAdd/vec_add_mpi.f90
program vec_add_mpi
    use mpi
    implicit none

    integer, parameter :: n = 10000000       ! total vector size
    double precision, allocatable :: a(:), b(:), c(:)
    double precision, allocatable :: local_a(:), local_b(:), local_c(:)
    integer :: ierr, rank, size
    integer :: i, chunk, start_idx, end_idx
    double precision :: t_start, t_end, elapsed_local, elapsed_max

    ! Initialize MPI
    call MPI_Init(ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, size, ierr)

    ! Determine chunk size per process
    chunk = n / size
    start_idx = rank * chunk + 1
    end_idx   = start_idx + chunk - 1

    ! Allocate local arrays
    allocate(local_a(chunk), local_b(chunk), local_c(chunk))

    ! Initialize local arrays
    local_a = 1.0d0
    local_b = dble([(i, i=start_idx,end_idx)])

    ! Synchronize processes before timing
    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    t_start = MPI_Wtime()

    ! Vector addition
    do i = 1, chunk
        local_c(i) = local_a(i) + local_b(i)
    end do

    ! Synchronize processes after computation
    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    t_end = MPI_Wtime()
    elapsed_local = t_end - t_start

    ! Reduce to get maximum elapsed time across all processes
    call MPI_Reduce(elapsed_local, elapsed_max, 1, MPI_DOUBLE_PRECISION, MPI_MAX, 0, MPI_COMM_WORLD, ierr)

    ! Allocate full vector on root process to gather results
    if (rank == 0) then
        allocate(a(n), b(n), c(n))
    end if

    ! Gather results from all processes
    call MPI_Gather(local_a, chunk, MPI_DOUBLE_PRECISION, a, chunk, MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)
    call MPI_Gather(local_b, chunk, MPI_DOUBLE_PRECISION, b, chunk, MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)
    call MPI_Gather(local_c, chunk, MPI_DOUBLE_PRECISION, c, chunk, MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)

    ! Print results from root
    if (rank == 0) then
        print *, "c[0] =", c(1), ", c[n-1] =", c(n), ", elapsed =", elapsed_max, "seconds"
    end if

    ! Clean up
    deallocate(local_a, local_b, local_c)
    if (rank == 0) then
        deallocate(a, b, c)
    end if

    call MPI_Finalize(ierr)
end program vec_add_mpi
